### ISSUE TITLE
grammar-constrained decoding: force valid recommendation forms from a small model

### ISSUE THREAD

**Issue by geisten** [geisten/geistshell#34]
The real unblock for real-model completion (and thus the LIFT half of #25/#26),
found empirically on the Pi.

## Why

Path B added an `(examples ...)` context slot (#32) so a small model gets
concrete worked recommendation forms. Measured against Gemma 4 E2B on the Pi:
even WITH exemplars, Gemma still terminates `budget` with
`SPG_RECOMMENDATION_REJECT_SYNTAX` — it cannot freely emit a valid
`(recommend ...)` s-expression. Few-shot is necessary but not sufficient
(docs/LEARNING.md).

geistagent solved exactly this class — a small model emitting a structured
call — with **grammar-constrained decoding**: the sampler is masked to only
the tokens the recommendation grammar allows, so the output is valid by
construction (BitNet/Gemma then produced valid tool calls). geistshell has no
such sampler mask; the model decodes freely and a 2B model rarely lands a
valid form.

## Scope

- **In:** a decoding mode that constrains the model's next-token choices to the
  `(recommend ...)` grammar (kinds, keys, value shapes) so every emitted form
  parses and matches the schema; opt-in, reusing the engine's sampler hook.
- **Out:** training/fine-tuning; changing the DSL; the free-decode path (stays
  default). Enum/argument value constraints beyond the form grammar are a
  follow-up.

## Acceptance criteria (done when)

- [ ] A constrained-decoding mode yields a valid, schema-matching recommendation
      form from Gemma 4 E2B on the Pi where free decoding produced a SYNTAX
      reject — measured, not assumed.
- [ ] Free decoding is unchanged and remains default.
- [ ] With it, a small model completes a simple shell task (verdict=pass) that
      it failed without it — the model-completable corpus (#25) becomes real.
- [ ] Then the LIFT experiments (#25 numerator, #26 benefit) can run: control
      vs geistshell, success rate + gain-per-token, on a model that can act.
- [ ] Unit-tested where the mask is pure; the real-model result is GGUF-gated.

## Note

This is the prerequisite the exemplar work (#32) revealed. Until it lands, the
learning loop's context-cost claim is proven (bench_context) but the
success-lift claim cannot be — every arm scores ~0 because the model cannot
emit the DSL.

## Technical pre-clarification (verified against geist v0.2.1 on the Pi)

**No engine fork is needed — geistshell can build this at its own layer.**
v0.2.1's `geist_util.h` already exposes the required hooks (all EXPERIMENTAL,
CPU-backend only):

- `geist_session_peek_logits(s, &n_vocab)` — pointer to the next-position
  logits + vocab size.
- `geist_session_prefill_tokens(s, n, tokens)` — advance the session with a
  CHOSEN token (no internal sampling).
- `geist_model_token_by_text(m, text)` / `geist_session_tokenize(...)` — text↔token.

Verified: the **transformer arch implements `peek_logits`**
(`src/archs/transformer/arch.c` `op_peek_logits` → `.peek_logits`), and Gemma is
a transformer — so Gemma on the Pi (CPU, transformer) supports it. The note in
the header that 'not every arch implements peek_logits (Mamba2 currently)' does
not affect us.

### Implementation shape

Today `spg_model_generate` (src/model/model_adapter.c) loops on
`geist_session_decode_step`, a black box that samples internally — no mask hook.
The constrained mode replaces that loop with:

    loop:
      logits = geist_session_peek_logits(s, &n_vocab)
      mask logits to the grammar-allowed token set (disallowed → -inf)
      token  = argmax(masked)              // greedy; sampling optional
      append token_to_str(token)
      geist_session_prefill_tokens(s, 1, &token)   // advance with the chosen token

So it is a new decode mode in the model adapter, opt-in, leaving
`decode_step` (free decode) as the default.

### The real work: a token-level grammar acceptor

The hard part is the grammar state machine: given the tokens emitted so far,
compute which next tokens keep the output a valid `(recommend ...)` form
(open paren, `recommend`, a valid `kind`, the required keys, value shapes,
close). **geistagent already did exactly this** — `agent.h`'s grammar-masked
decode (NAME via constrained decode, args via schema) is the reference. Porting
that acceptor is the bulk of the effort; the engine plumbing above is small.

### Caveats to carry

- The hooks are EXPERIMENTAL — a minor engine bump may change signatures; pin
  and re-verify on `make update-engine`.
- CPU-backend only (the Pi is CPU — fine; a future GPU backend would not have it).
- Byte-vs-token boundary: the mask operates on tokens, but the grammar is over
  characters — the acceptor must handle multi-char tokens and partial forms
  (the same problem agent.h solved with per-alphanumeric-run matching).

**Comment by geisten** [comment]
**First cut merged (#35): forced-prefix constrained decoding.**

Force the output to begin with `(recommend (kind `, then decode freely — via geist v0.2.1 `geist_util.h` (tokenize + prefill_tokens), no engine fork. `agent --constrained`.

Pi validation (Gemma 4 E2B) — the reject reason **moved forward**:
- free decode / exemplars only → `REJECT_SYNTAX` (not even parseable)
- `--constrained` → `REJECT_UNKNOWN_KIND` — parses as a recommendation up to the kind field; Gemma only picks an invalid kind name after the forced opening.

The mechanism is proven and the exact next constraint is now known.

**Remaining in #34 — the token-level acceptor:**
1. Constrain the KIND token to the valid enum (peek_logits + a mask over the kind names) — expected to yield the first fully valid form. This is the smallest next step and the highest-value one.
2. Generalize the mask past the kind field to the rest of the grammar (geistagent `agent.h` as the reference acceptor).

Leaving #34 open for the acceptor work.

**Comment by geisten** [comment]
**Constrained decoder complete (stages 1, 2, 2b merged: #36, #37, #38).**

Reject/deny chain fully resolved on the Pi against Gemma 4 E2B:
`SYNTAX` (free) → `UNKNOWN_KIND` (forced prefix) → `MISSING_FIELD` (kind mask) → **valid form** (field scaffold) → `denied` (free capability) → **allowed + executed** (capability mask).

Final run:
```
(recommend (kind simulator) (capability "sim.act") (cost 1) (uses_network false) (confidence_bp 5000) (reason "Simulate a scenario"))
(policy_decision (decision allow) (deny_reason SPG_POLICY_DENY_NONE) ...)
(sim_result (action patch_vulnerability) (mutated true) (risk_before 35300) (risk_after 32100))
```
The simulator ran and dropped risk — real effect from a model that could not emit the DSL at all before.

**Acceptance criteria:**
- [x] valid, schema-matching form where free decode gave SYNTAX — measured.
- [x] free decode unchanged and default.
- [x] the model now acts: policy allows the masked form and the action executes (`sim_result`, risk 35300→32100). A clean `verdict=pass` needs an `expect` corpus — that is the LIFT stage.
- [ ] LIFT experiments (#25 numerator, #26 benefit): control vs constrained on an `expect` corpus — **now unblocked**, tracked as stage 3.
- [x] unit-tested where pure: `test_grammar_mask` (choice mask + every kind's scaffold parsed through `recommendation.c`, no GGUF).

Approach shipped: forced prefix + kind mask + per-kind field scaffold + capability mask — the pragmatic scaffold, not a general token-level acceptor (that stays the follow-up if the DSL ever becomes open/extensible). Leaving #34 open for the LIFT measurement.

**Comment by geisten** [comment]
**Stage 3 — LIFT measured (#39 merged).** The #25 numerator, on the model that could not act before.

`eval/bench_lift.sh`, four scenario variants × two arms, real Gemma 4 E2B on the Pi (~18 min):

| | free valid | constr valid | free acted | constr acted | done |
|---|---|---|---|---|---|
| **4 tasks** | 0/4 | **4/4** | 0/4 | **4/4** | 0/4 both |

- **valid** = policy-allowed form emitted; **acted** = `sim_result` (an action executed, world moved); **done** = verdict=pass.
- Constrained decoding: **0→100%** valid + executing actions. Free decode never lands the DSL.
- `done` 0/4 both arms, honestly: the CLI expect judge requires `termination=FINISHED`, and greedy Gemma keeps acting rather than emitting `(kind finish)`. Acting validly vs choosing to stop are separate behaviours — the decoder fixed the first.

All #34 acceptance criteria met (valid form where free gave SYNTAX; free decode unchanged/default; model now acts — valid+allowed+executed; unit-tested where pure; LIFT measured). Closing #34 as the constrained decoder is complete and its lift is quantified.

Follow-on (separate issues): the **finish/termination** behaviour (a nudge once risk is under threshold, or a terminating corpus) to convert `acted` into `done=verdict=pass`; and the full token-level grammar acceptor if the DSL ever becomes open/extensible.
